src/Microsoft.DotNet.ProjectModel/Workspace/WorkspaceContext.cs

```diff
@@ -0,0 +1,328 @@
+                var builder = new ProjectContextBuilder()
+                    .WithProjectResolver(path => GetProject(path))
+                    .WithLockFileResolver(path => GetLockFile(path))
+                    .WithProject(project)
+                    .WithTargetFramework(framework);
+
+                currentEntry.ProjectContext = builder.Build();
+            }
+
+            return currentEntry;
+        }
+
+        private ProjectContextEntry AddProjectContextEntryWithDependency(string projectDirectory,
+                                                                         NuGetFramework framework,
+                                                                         ProjectContextEntry currentEntry)
+        {
+            var entry = AddProjectContextEntry(projectDirectory, framework, currentEntry);
+            if (!entry.HasDependencyResolved)
+            {
+                entry.ResolveDependencies(Configuration);
+            }
+
+            return entry;
+        }
+
+        private class FileModelEntry<TModel> where TModel : class
+        {
+            private DateTime _lastWriteTime;
+
+            public TModel Model { get; set; }
+
+            public string FilePath { get; set; }
+
+            public void UpdateLastWriteTime()
+            {
+                _lastWriteTime = File.GetLastWriteTime(FilePath);
+            }
+
+            public bool IsInvalid
+            {
+                get
+                {
+                    if (Model == null)
+                    {
+                        return true;
+                    }
+
+                    if (!File.Exists(FilePath))
+                    {
+                        return true;
+                    }
+
+                    return _lastWriteTime < File.GetLastWriteTime(FilePath);
+                }
+            }
+
+            public void Reset()
+            {
+                Model = null;
+                FilePath = null;
+                _lastWriteTime = DateTime.MinValue;
+            }
+        }
+
+        private static List<string> PathResolve(string projectPath)
+        {
+            if (File.Exists(projectPath))
+            {
+                var filename = Path.GetFileName(projectPath);
+                if (!Project.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase) &&
+                    !GlobalSettings.FileName.Equals(filename, StringComparison.OrdinalIgnoreCase))
+                {
+                    return null;
+                }
+
+                projectPath = Path.GetDirectoryName(projectPath);
+            }
+
+            if (File.Exists(Path.Combine(projectPath, Project.FileName)))
+            {
+                return new List<string> { projectPath };
+            }
+
+            if (File.Exists(Path.Combine(projectPath, GlobalSettings.FileName)))
+            {
+                var root = ProjectRootResolver.ResolveRootDirectory(projectPath);
+                GlobalSettings globalSettings;
```

Comment:
`WorkspaceContext` should probably contain the full `GlobalSettings` object rather than just the project paths. My thought is that `WorkspaceContext` can actually **replace** `GlobalSettings`, or at least fully encapsulate it.